Proteotyping
============
Proteotyping uses peptides generated by mass spectrometry proteomics to:

* identify and estimate the taxonomic composition of a microbial sample.
* detect expressed antibiotic resistance proteins.

Proteotyping does not require a priori information about the contents of a
sample and is suitable for analysis of both pure cultures, mixed samples, and
clinical samples.
Have a look in the online `documentation`_ to learn how to use Proteotyping.

.. _documentation: https://proteotyping.readthedocs.org


About
*****
:Authors: Fredrik Boulund
:Contact: fredrik.boulund@chalmers.se
:License: BSD


Installation 
************
Detailed installation instructions are available in the online
`documentation`_.

Download and install Anaconda Python 3.5 and create a conda environment
using the following commands::

    $ wget https://bitbucket.org/chalmersmathbioinformatics/proteotyping/environment.yml
    $ conda env create -f environment.yml

This will create a conda environment called ``proteotyping`` that contains 
all the required dependencies, and the ``proteotyping`` package itself. 

Dependencies
------------
Proteotyping depends on the following Python packages, easily installable via
``conda`` and ``pip``.

 * `ETE Toolkit`_ (ete3)
 * `XlsxWriter`_ (xlsxwriter)

.. _XlsxWriter: http://xlsxwriter.readthedocs.org/
.. _ETE Toolkit: http://etetoolkit.org/

Running
*******
In order to run Proteotyping, some databases need to be prepared. Please
consult the online `documentation`_ for details on how to build these.

The input for Proteotyping is mappings of peptides to reference genomes in
blast8 tabular format. We recommend using `BLAT`_ or `pBLAT`_ for mapping.

.. _BLAT: https://genome.ucsc.edu/FAQ/FAQblat.html
.. _pBLAT: http://icebert.github.io/pblat/